File: src/main/kotlin/com/company/lolapi/ApiRequestLimiter.java
package com.company.lolapi;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.function.Supplier;

public class ApiRequestLimiter implements AutoCloseable {
    private final List<RequestsRateLimit> limits;

    protected ApiRequestLimiter(List<RequestsRateLimit> limits) {
        this.limits = Collections.unmodifiableList(limits);
    }

    public static ApiRequestLimiterBuilder builder() {
        return new ApiRequestLimiterBuilder();
    }

    public void execute(Runnable runnable) {
        acquire();
        runnable.run();
    }

    public <T> T execute(Supplier<T> supplier) {
        acquire();
        return supplier.get();
    }

    public void acquire() {
        for (RequestsRateLimit limit : limits) {
            limit.acquire();
        }
    }

    @Override
    public void close() {
        for (RequestsRateLimit limit : limits) {
            limit.shutdown();
        }
    }

    public static class ApiRequestLimiterBuilder {
        private List<RequestsRateLimit> limits;

        protected ApiRequestLimiterBuilder() {
            limits = new ArrayList<>();
        }

        public ApiRequestLimiterBuilder add(int requestsCount, TimeUnit timeUnit, int unitsCount) {
            return add(requestsCount, timeUnit, unitsCount, true);
        }

        public ApiRequestLimiterBuilder add(int requestsCount, TimeUnit timeUnit, int unitsCount, boolean equable) {
            RequestsRateLimit limit = equable
                    ? new RequestsRateLimitEquable(requestsCount, timeUnit, unitsCount)
                    : new RequestsRateLimitNonEquable(requestsCount, timeUnit, unitsCount);
            limits.add(limit);
            return this;
        }

        public ApiRequestLimiter build() {
            Collections.sort(limits);
            return new ApiRequestLimiter(limits);
        }
    }
}

File: src/main/kotlin/com/company/lolapi/ApiRequestLimiter.kt
package com.company.lolapi

import java.util.concurrent.TimeUnit
import java.util.function.Supplier

class ApiRequestLimiter private constructor(private val limits: List<RequestsRateLimit>) : AutoCloseable {

    fun execute(runnable: Runnable) {
        acquire()
        runnable.run()
    }

    fun <T> execute(supplier: Supplier<T>): T {
        acquire()
        return supplier.get()
    }

    fun acquire() {
        for (limit in limits) {
            limit.acquire()
        }
    }

    override fun close() {
        for (limit in limits) {
            limit.shutdown()
        }
    }

    class ApiRequestLimiterBuilder internal constructor() {
        private val limits = mutableListOf<RequestsRateLimit>()

        fun add(requestsCount: Int, timeUnit: TimeUnit, unitsCount: Int, equable: Boolean = true): ApiRequestLimiterBuilder {
            limits.add(
                    if (equable) RequestsRateLimit(requestsCount, timeUnit, unitsCount)
                    else RequestsRateLimitNonEquable(requestsCount, timeUnit, unitsCount))
            return this
        }

        fun build(): ApiRequestLimiter {
            limits.sort()
            return ApiRequestLimiter(limits)
        }
    }

    companion object {
        fun builder(): ApiRequestLimiterBuilder {
            return ApiRequestLimiterBuilder()
        }
    }
}
